src/timer.c

```diff
@@ -140,31 +190,56 @@ static int __proc_timers(timer_link_t * head)
  */
 static void timer_callback(void)
 {
-	if (TI_timers)
-	{
-		while (__proc_timers(TI_timers)) ;
-	}
-	else
-	{
-		int now = TICKS_READ();
-		total_ticks += now;
-		write_count(0);
-		write_compare(0x7FFFFFFF);
-	}
+	while (__proc_timers(TI_timers))
+		{}
+
+	// Update counter for next interrupt.
+	timer_update_compare(TI_timers);
+}
+
+/**
+ * @brief Timer callback overflow function
+ *
+ * This function is the callback of the internal overflow timer, which
+ * is configured by timer_init() and is used to create a 64-bit timer
+ * accessed by timer_ticks().
+ */
+static void timer_callback_overflow(int ovfl)
+{
+	ticks64_high++;
 }
 
 /**
  * @brief Initialize the timer subsystem
  *
- * @note This will currently mess with get_ticks, get_ticks_ms and related wait
- * routines as it is explicitly manipulating the system timer.
+ * This function will reset the COP0 ticks counter to 0. Even if you
+ * later access the hardware counter directly (via TICKS_READ()), it should not
+ * be a problem if you call timer_init() early in the application main.
  *
+ * Do not modify the COP0 ticks counter after calling this function. Doing so
+ * will impede functionality of the timer module.
  */
 void timer_init(void)
 {
-	total_ticks = 0;
+	/* Create first timer for overflows: expires when counter is 0 and
+	 * has a period of 2**32. */
+	timer_link_t *timer = malloc(sizeof(timer_link_t));
+	if (timer)
+	{
+		timer->left = 0;
+		timer->set = 0;
+		timer->flags = TF_CONTINUOUS | TF_OVERFLOW;
+		timer->callback = timer_callback_overflow;
+		timer->next = NULL;
+
+		TI_timers = timer;
+	}
+
+	ticks64_high = 0;
 	write_count(0);
-	write_compare(0x7FFFFFFF);
+	write_compare(0);
```

Comment:
No there's no need to trigger an interrupt because we're just resetting. The sequence does not trigger an interrupt because first count is written, and then compare, and by the time compare is set to 0, count is at least 1. Now that you mention, it's probably a little "edgy" and can also confuse emulators. Maybe we should just set count to 1 and keep compare to 0.

I don't get your comment about handling the overflow at FFFFFFFF instead. Can you elaborate on why you think it's necessary?

Reply:
I didn't get why count is 1 when we write the compare. It is being updated at half maximum instruction rate, so I expect it to not being incremented if we are on the other half. This was what I'm thinking;
```
<-- icrement was on this cycle
write_count(0); # <-- we reset to 0
write_compare(0); # <-- we have written compare but it is now 1 and we're safe
```
The above case is safe. But what if;
```
write_count(0); # <-- icrement was on this cycle, but we reset to 0
write_compare(0); # <-- we have written compare while count was 0, will this trigger an interrupt?
```

If the interrupt only fires when count is incremented, this should not be an issue, but if it is fired when they are equal, this maybe a problem. I couldn't find more info on the exact behavior.

Reply:
Im not even sure on the first case now as I dont know the exact operation (is count incremented before we write compare?). If you are sure it wont trigger this is ok.

Reply:
Also what is the initial value for compare? What if we -by chance- are setting count to it? Maybe we should disable ints, set count to 1 compare to 0 and enable. This will make sure we dont trigger it. Even if initial compare is 1, ints are disabled and we write compare next which will clear the interrupt bit before we re-enable.

Reply:
One last thing: we can only enable timer interrupt on timer init instead of init interrupts so that we dont cause unnecessary interrupts.

Reply:
Done.